java/vortex-spark/src/main/java/dev/vortex/spark/read/PrefetchingIterator.java

```diff
@@ -0,0 +1,97 @@
+/**
+ * (c) Copyright 2025 SpiralDB Inc. All rights reserved.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.vortex.spark.read;
+
+import java.util.Iterator;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.ToLongFunction;
+
+final class PrefetchingIterator<T> implements Iterator<T>, AutoCloseable {
+    private final BlockingQueue<T> fetched = new LinkedBlockingQueue<>();
+    private final Thread producerThread;
+    private final Iterator<T> delegate;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final AtomicLong bufferBytes = new AtomicLong(0);
+    private final long maxBufferSize;
+    private final ToLongFunction<T> sizeFunc;
+
+    PrefetchingIterator(Iterator<T> delegate, long maxBufferSize, ToLongFunction<T> sizeFunc) {
+        this.delegate = delegate;
+        this.maxBufferSize = maxBufferSize;
+        this.sizeFunc = sizeFunc;
+        this.producerThread = new Thread(this::prefetchLoop, "vortex-prefetch-thread");
+        producerThread.setDaemon(true);
+        producerThread.start();
+    }
+
+    private void prefetchLoop() {
+        try {
+            while (!closed.get() && delegate.hasNext()) {
+                while (bufferBytes.get() > maxBufferSize) {
+                    // TODO(aduffy): avoid spin waiting
```

Comment:
I gave a brief think on this and it was not immediately obvious to me how to do this.

Would it be better for us to implement a bounded blocking queue of elements for now and think about how to implement something in terms of byte size without a spin-wait? I'm a bit worried this is the kind of code one merges and forgets to come back to.

Reply:
I think this is better now. I changed it from a sleep to instead using an intrinsic monitor condvar that is notified when an element is dequeued and waited in the producer here